### ISSUE TITLE
ssh-keygen is not included with Windows Git

### ISSUE THREAD

**Issue by cmerris-inmoment** [progit/progit2#1090]
https://git-scm.com/book/en/v2/Git-on-the-Server-Generating-Your-SSH-Public-Key

Current documentation states that "ssh-keygen, which is provided with the SSH package on Linux/Mac systems and comes with Git for Windows:"

This is no longer true - I have tried installed Git for Windows twice today and once installation is complete, running "ssh-keygen" from a command prompt gives an error - since the tool is not on the PATH and cannot be found.  I can't find it anywhere in the Git install directory either.

**Comment by cmerris-inmoment** [comment]
My git version - is the current windows git version for today: 2.17.1.2 (executable is named Git-2.17.1.2-64-bit.exe).  I'm also on Windows 10.

**Comment by ben** [comment]
Oh no! Do you have any idea how one could go from a bare Git for Windows install to having a working ssh-keygen? Does MinGW come with a package manager?

**Comment by cmerris-inmoment** [comment]
My windows install doesn't include a "C:\Windows\System32\OpenSSH" folder and ssh-keygen is not on the path.

It's irrelevant to my original issue, however.  The Git Book states "...you can create them by running a program called ssh-keygen, which is provided with the SSH package on Linux/Mac systems and comes with Git for Windows..."

This is untrue.  ssh-keygen does not come with Git for Windows.  I can't find it anywhere in my git installation directory.  The documentation is incorrect.

**Comment by cmerris-inmoment** [comment]
> Oh no! Do you have any idea how one could go from a bare Git for Windows install to having a working ssh-keygen? Does MinGW come with a package manager?

@ben I installed Git on Windows from git-scm.com (the same site with the book I've quoted).  No package managers were involved in my installation.  Even if I was able to use a package manager to install ssh-keygen, it wouldn't change the fact that ssh-keygen is not included with Git for Windows.

I didn't create the issue to get help with ssh-keygen - I created the issue to point out the book's error.

**Comment by WebMechanic** [comment]
> This is untrue. ssh-keygen does not come with Git for Windows. I can't find it anywhere in my git installation directory. The documentation is incorrect.

Maybe you ticked a different checkbox during Git setup or missed one, but this is what I have in my Git folder:
```
C:\Program Files\Git>dir ssh*.exe /B /S
C:\Program Files\Git\usr\bin\ssh.exe
C:\Program Files\Git\usr\bin\ssh-add.exe
C:\Program Files\Git\usr\bin\ssh-agent.exe
C:\Program Files\Git\usr\bin\sshd.exe
C:\Program Files\Git\usr\bin\ssh-keygen.exe
C:\Program Files\Git\usr\bin\ssh-keyscan.exe
C:\Program Files\Git\usr\bin\ssh-pageant.exe
C:\Program Files\Git\usr\lib\ssh\ssh-keysign.exe
C:\Program Files\Git\usr\lib\ssh\ssh-pkcs11-helper.exe
```
That's `git version 2.19.1.windows.1` for you.

However: The one in the system folder appear to be "native" windows apps, those in Git/usr/bin are based on MinGW.

My point was that it appears that "_meanwhile_" OpenSSH tools come bundled with **Win10 Professional 1803**+ in the aforementioned system folder. The **Pro** version might do the difference -- or enabling the "Developer Extensions".
The documentation could be updated to reflect/include that change.
It's certainly not wrong stating that ssh-keygen.exe comes with Git for Windows (as well).

**Comment by cmerris-inmoment** [comment]
Thanks for helping me find it.  If I had known where to look for ssh-keygen I wouldn't have created this ticket.

I feel it would be helpful to update the book to alleviate this confusion for Windows users in the future - but it's up to you guys.

**Comment by cmerris-inmoment** [comment]
.

**Comment by WebMechanic** [comment]
For the record: MSDN blog [What’s new for the Command Line in Windows 10 version 1803](https://blogs.msdn.microsoft.com/commandline/2018/03/07/windows10v1803/)

> The SSH client and key agent are enabled and available by default in this Update, and the SSH server is an optional feature-on-demand. 

The post doesn't mention any specific Win10-flavour (Basic/Home/Pro/Server/Enterprise) so I presume OpenSSH is available as of Service Release March 2018 (v10.1803).

**Comment by jhabdas** [comment]
Is it included with [Cmder](http://cmder.net)?

**Comment by jnavila** [comment]
@WebMechanic With these files in this place , is it available in git-bash and Powershell?

**Comment by WebMechanic** [comment]
TL;DR: either way works.

> @WebMechanic With these files in this place , is it available in git-bash and Powershell?

By default git-bash will call it's "local" (MinGW) version and the Windows CLIs will use the "native" ones from the system folder.

git-bash
```
$ which ssh-keygen
/usr/bin/ssh-keygen
```

What comes with "Windows 10 version 1803"  (C:\Windows\System32\OpenSSH\) will run with both Powershell and cmd.exe respectively and will take precendes (PATH) when called from there:
```
D:\>where ssh-keygen
C:\Windows\System32\OpenSSH\ssh-keygen.exe
```

That would be your defaults, but you can use either version if you wish so in **git-bash** or **cmd.exe** (or if PATH was modified during Git setup to include the provided "Gnu tools"). Powershell however doesn't seem to care much about MinGW executables and does nothing if you attempt to run one - there's no output or error whatsoever.

git-bash calling a Win10 executable:
```
$ /c/Windows/System32/OpenSSH/ssh-keygen.exe
Generating public/private rsa key pair.
Enter file in which to save the key ...
```

Win10 **cmd.exe** calling MinGW (git-bash) executable:
```
D:\>"C:\Program Files\Git\usr\bin\ssh-keygen.exe"
Generating public/private rsa key pair.
Enter file in which to save the key ...
```

HTH.

**Comment by AliFahed** [comment]
Hey everyone, I just encounter similar issue, in my PowerShell ssh was not defined.

How I solved it:
Go to Settings --> Optional features, In Installed features section search for OpenSSH Client if you found it uninstall it and if not it's ok, in both cases in the same page click Add a feature and search for OpenSSH Client and finally install it.

Go to PowerShell check the result by using: ssh

**Comment by WorikQCI** [comment]
FYI I have a brand new windows installation adn downloaded git

No `ssh-keygen` in gitbash but is in windows powershell.

**Comment by WorikQCI** [comment]
I am unsure that "safely say" is a useful concept in this context.

I have not been able to make the Windows Power Shell ssh stuff do what I want it to do, and what is instructed.  

....ongoing
